pkg/ha/source.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ha
+
+import (
+	"context"
+	"fmt"
+
+	"go.etcd.io/etcd/clientv3"
+
+	"github.com/pingcap/dm/dm/common"
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/pkg/etcdutil"
+)
+
+// PutSourceCfg puts the config of the upstream source into etcd.
+// k/v: sourceID -> source config.
+func PutSourceCfg(cli *clientv3.Client, cfg config.MysqlConfig) (int64, error) {
+	value, err := cfg.Toml()
+	if err != nil {
+		return 0, err
+	}
+	key := common.UpstreamConfigKeyAdapter.Encode(cfg.SourceID)
+
+	return etcdutil.DoOpsInOneTxn(cli, clientv3.OpPut(key, value))
+}
+
+// GetSourceCfg gets the config of the specified source.
+// if the config for the source not exist, return with `err == nil` and `revision=0`.
+func GetSourceCfg(cli *clientv3.Client, source string, rev int64) (config.MysqlConfig, int64, error) {
```

Comment:
why need revision in the argument?

Reply:
After worker is started, worker will get and watch source bound stage which will returns a revision. We use revision to get relative source config for a source bound event.
For example:
the time sequence is:
1. Put sourceCfg
2. Put sourceBound
3. Delete sourceBound
4. Delete sourceCfg
5. Put sourceCfg2(but has same sourceID with sourceCfg)

If when worker get newest sourceCfg at step.2 etcd has done step1~5, worker will start source with the config of sourceCfg2 which is not correct.